rfcs/42-world-cell-queries.md

```diff
@@ -0,0 +1,236 @@
+# Feature Name: `world_cell_queries`
+
+## Summary
+
+Allow component queries to be issued through `WorldCell` API in single-threaded systems.
+
+Allow the user to relax usual borrowing semantics in favour of runtime checks. Allow for simultaneous querying
+of world data that is hard to prove disjoint at compile time. Rely on runtime assertions to provide safety.
+
+Allow performing world modifications through WorldCell shared access using command API.
+
+Make the world cell queries always reflect the most recent updated world data, without requiring explicit
+sync points inside the single-threaded systems.
+
+## Motivation
+
+Bevy ECS is currently quite well suited for high data throughput and relatively low complexity scenarios.
+Most focus is being put on general performance and ergonomics of safe multi-threaded query system.
+That high performance API imposes some hard limitations on how to structure the data, effectively requiring
+every system to be carefully designed around rust borrowing rules, as well as making complex data dependency chains
+sometimes very hard to deal with.
+
+While this trade-off might be right for a lot of scenarios, there is currently no way out of it. Most notably, it can
+easily get in the way of complex gameplay logic prototyping, where development speed is much more important than raw performance.
+Similarly, some systems might not be performance critical, but annoying to write with usual query limitations in place.
+
+The trade-off of performance versus code complexity should be a decision that a developer could make on their own.
+
+## User-facing explanation
+
+### World cell query
+
+When having access to a `WorldCell`, you are free to issue any ECS queries you wish, at any time. Even ones that seemingly break rust principles.
+
+```rust
+pub fn single_threded_system(world: &mut World) {
+    let mut cell = world.cell();
+
+    // create two seemingly conflicting queries and iterate over them
+    let query_ref = cell.query::<(Entity, &Transform)>();
+    let query_mut = cell.query::<(Entity, &mut Transform)>();
+    let mut iter_ref = query_ref.iter();
+    let mut iter_mut = query_mut.iter();
+
+    iter_ref.next(); // offset mutable query by one item to avoid aliasing
```

Comment:
It will still alias. All items returned by an iterator are allowed to exist at the same time. For example when doing `.collect()`.

Reply:
Then the iterator has to return `std::cell::Ref` or `std::cell::RefMut`. You can't return the inner reference as once the `Ref` or `RefMut` is dropped, you are no longet allowed to access the inner value, as someone else can now borrow it.

Reply:
I was confused by this too, RFC should explicitly mention that the iterators return a `Ref`/`RefMut`/some other lock guard

Reply:
It is mentioned, but in the implementation details. Maybe that's not enough. 

> Instead of usual component references, queries return all component data through wrappers types. Similar to already existing Mut wrapper for &mut X world fetches. The existance of those wrappers allows for both runtime checks and "world snapshots" with all commands applied - the wrapper can reference either world or a command queue data.